### PR TITLE
Add UWP support

### LINKED ISSUE #1879: Unable to install package on UWP
Hi!

**Type**: Bug
**NLog version**: 5.0.0-beta04 (Also for previous versions)
**Platform**: .UWP (Win10)

I can't install the package on my project, I have this errors: Package restore failed. Rolling back package changes for 'App2'.

**Error detail**:

```
System.Threading.AccessControl 4.0.0 provides a compile-time reference assembly for System.Threading.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm.
System.Security.AccessControl 4.0.0 provides a compile-time reference assembly for System.Security.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm.
System.Threading.ThreadPool 4.0.10 provides a compile-time reference assembly for System.Threading.ThreadPool on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm.
System.Threading.Thread 4.0.0 provides a compile-time reference assembly for System.Threading.Thread on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm.
System.IO.FileSystem.AccessControl 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm.
System.IO.FileSystem.Watcher 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.Watcher on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm.
One or more packages are incompatible with UAP,Version=v10.0 (win10-arm).
System.Threading.AccessControl 4.0.0 provides a compile-time reference assembly for System.Threading.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm-aot.
System.Security.AccessControl 4.0.0 provides a compile-time reference assembly for System.Security.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm-aot.
System.Threading.ThreadPool 4.0.10 provides a compile-time reference assembly for System.Threading.ThreadPool on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm-aot.
System.Threading.Thread 4.0.0 provides a compile-time reference assembly for System.Threading.Thread on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm-aot.
System.IO.FileSystem.AccessControl 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm-aot.
System.IO.FileSystem.Watcher 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.Watcher on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm-aot.
One or more packages are incompatible with UAP,Version=v10.0 (win10-arm-aot).
System.Threading.AccessControl 4.0.0 provides a compile-time reference assembly for System.Threading.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x64.
System.Security.AccessControl 4.0.0 provides a compile-time reference assembly for System.Security.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x64.
System.Threading.ThreadPool 4.0.10 provides a compile-time reference assembly for System.Threading.ThreadPool on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x64.
System.Threading.Thread 4.0.0 provides a compile-time reference assembly for System.Threading.Thread on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x64.
System.IO.FileSystem.AccessControl 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x64.
System.IO.FileSystem.Watcher 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.Watcher on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x64.
One or more packages are incompatible with UAP,Version=v10.0 (win10-x64).
System.Threading.AccessControl 4.0.0 provides a compile-time reference assembly for System.Threading.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x64-aot.
System.Security.AccessControl 4.0.0 provides a compile-time reference assembly for System.Security.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x64-aot.
System.Threading.ThreadPool 4.0.10 provides a compile-time reference assembly for System.Threading.ThreadPool on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x64-aot.
System.Threading.Thread 4.0.0 provides a compile-time reference assembly for System.Threading.Thread on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x64-aot.
System.IO.FileSystem.AccessControl 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x64-aot.
System.IO.FileSystem.Watcher 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.Watcher on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x64-aot.
One or more packages are incompatible with UAP,Version=v10.0 (win10-x64-aot).
System.Threading.AccessControl 4.0.0 provides a compile-time reference assembly for System.Threading.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x86.
System.Security.AccessControl 4.0.0 provides a compile-time reference assembly for System.Security.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x86.
System.Threading.ThreadPool 4.0.10 provides a compile-time reference assembly for System.Threading.ThreadPool on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x86.
System.Threading.Thread 4.0.0 provides a compile-time reference assembly for System.Threading.Thread on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x86.
System.IO.FileSystem.AccessControl 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x86.
System.IO.FileSystem.Watcher 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.Watcher on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x86.
One or more packages are incompatible with UAP,Version=v10.0 (win10-x86).
System.Threading.AccessControl 4.0.0 provides a compile-time reference assembly for System.Threading.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x86-aot.
System.Security.AccessControl 4.0.0 provides a compile-time reference assembly for System.Security.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x86-aot.
System.Threading.ThreadPool 4.0.10 provides a compile-time reference assembly for System.Threading.ThreadPool on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x86-aot.
System.Threading.Thread 4.0.0 provides a compile-time reference assembly for System.Threading.Thread on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x86-aot.
System.IO.FileSystem.AccessControl 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.AccessControl on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x86-aot.
System.IO.FileSystem.Watcher 4.0.0 provides a compile-time reference assembly for System.IO.FileSystem.Watcher on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-x86-aot.
One or more packages are incompatible with UAP,Version=v10.0 (win10-x86-aot).
```

Thank you in advance.

### PR DESCRIPTION
**Pull request by gtbuchanan**
Fixes #1879. I have not tested thoroughly since I really only need basic console logging functionality right now, but this got it to build and log successfully in a UWP project.

### CODE REVIEW

**Comment by 304NotModified** [comment]
great!

UWP and unit tests from console, any idea of that could work?

File: src/NLog/project.json [hunk 1]
```diff
@@ -221,9 +221,30 @@
...
+        "System.Net.Requests": "4.3.0",
+        "System.Reflection.TypeExtensions": "4.3.0",
+        "System.Runtime.Serialization.Primitives": "4.3.0",
+        "System.Security.Principal": "4.3.0",
+        "System.Threading": "4.3.0",
+        "System.Xml.ReaderWriter": "4.3.0",
+        "System.Xml.XmlDocument": "4.3.0"
+      },
+      "buildOptions": {
+        "define": [ "NETSTANDARD", "NETSTANDARD1_4", "WINDOWS_UWP" ]
```
Comment (304NotModified) [inline src/NLog/project.json]: Why do we need `NETSTANDARD1_4`?

**Comment by snakefoot** [comment]
Instead of introducing NETSTANDARD1_4 (Initial version to support uap10.0). Why not use NETSTANDARD1_5?

Any reason for using use Task.Delay to perform Thread.Sleep ? Not possible to depend on System.Threading.Thread? If necessary can it be put in a helper method?

Why not change ThreadLocalStorageHelper to use #elif NETSTANDARD instead of #elif  WINDOWS_UWP ? Would be nice to only have NETSTANDARD-versions defined in NLog, and not  WINDOWS_UWP.

**Comment by snakefoot** [comment]
Can understand that Thread-library is not available for UWP-apps until NetStandard2_0:

Maybe something like this could be used:

```
        internal static void StartAsyncTask(Action<object> waitCallback, object state)
        {
#if NETSTANDARD
            System.Threading.Tasks.Task.Factory.StartNew(waitCallback, state);
#else
            ThreadPool.QueueUserWorkItem(new WaitCallback(waitCallback), state);
#endif
        }

        internal static void WaitForDelay(TimeSpan delay)
        {
#if NETSTANDARD
            System.Threading.Tasks.Task.Delay(delay).Wait();
#else
            Thread.Sleep(delay);
#endif
        }
```

File: src/NLog/project.json (hunk 1 above)
Comment (gtbuchanan) [inline src/NLog/project.json]: It's not used, so it can be removed. I only added it because UWP only seems to support NET Standard 1.4.

**Comment by gtbuchanan** [comment]
@304NotModified I'm not sure about the unit tests as I haven't tried. I'm new to UWP myself.

@snakefoot The NETSTANDARD1_4 flag is not used, so it can be removed. You are correct about `Thread.Sleep`. UWP does not support most of the old `System.Threading` namespace. Pretty much only `Task`.

My reason for using the WINDOWS_UWP flag was to minimize the changes for other platforms. You may very well be able to use NETSTANDARD instead. I agree the common logic should be moved to helper methods.

Unfortunately, I don't have the capacity right now to work on this any further. I only made the changes necessary to get the build passing so I could have basic functionality for my employer's new Windows app. I simply wanted to show support is possible with fairly minimal effort.

**Comment by 304NotModified** [comment]
@gtbuchanan the netstandard1.3 version isn't working for you? It should support UWP. See https://docs.microsoft.com/en-us/dotnet/articles/standard/library

**Comment by gtbuchanan** [comment]
@304NotModified No. It contains libraries that are not supported in UWP (as mentioned [here](https://github.com/NLog/NLog/pull/2055#issuecomment-292736800) and #1879).

**Comment by 304NotModified** [comment]
superseded by https://github.com/NLog/NLog/pull/2055

thanks @gtbuchanan !

Will release nlog5 beta07 soon (probably today)

**Pull request closed without merging**